source/common/router/rds_impl.cc

```diff
@@ -51,110 +52,62 @@ RouteConfigProviderSharedPtr RouteConfigProviderUtil::create(
-                          [this]() { local_init_manager_.initialize(local_init_watcher_); }),
-      local_init_watcher_(fmt::format("RDS local-init-watcher {}", rds.route_config_name()),
-                          [this]() { parent_init_target_.ready(); }),
-      local_init_target_(
-          fmt::format("RdsRouteConfigSubscription local-init-target {}", route_config_name_),
-          [this]() { subscription_->start({route_config_name_}); }),
-      local_init_manager_(fmt::format("RDS local-init-manager {}", route_config_name_)),
-      stat_prefix_(stat_prefix),
-      stats_({ALL_RDS_STATS(POOL_COUNTER(*scope_), POOL_GAUGE(*scope_))}),
-      route_config_provider_manager_(route_config_provider_manager),
-      manager_identifier_(manager_identifier), optional_http_filters_(optional_http_filters) {
-  const auto resource_name = getResourceName();
-  subscription_ =
-      factory_context.clusterManager().subscriptionFactory().subscriptionFromConfigSource(
-          rds.config_source(), Grpc::Common::typeUrl(resource_name), *scope_, *this,
-          resource_decoder_, {});
-  local_init_manager_.add(local_init_target_);
-  config_update_info_ =
-      std::make_unique<RouteConfigUpdateReceiverImpl>(factory_context, optional_http_filters_);
-}
-
-RdsRouteConfigSubscription::~RdsRouteConfigSubscription() {
-  // If we get destroyed during initialization, make sure we signal that we "initialized".
-  local_init_target_.ready();
-
-  // The ownership of RdsRouteConfigProviderImpl is shared among all HttpConnectionManagers that
-  // hold a shared_ptr to it. The RouteConfigProviderManager holds weak_ptrs to the
-  // RdsRouteConfigProviders. Therefore, the map entry for the RdsRouteConfigProvider has to get
-  // cleaned by the RdsRouteConfigProvider's destructor.
-  route_config_provider_manager_.dynamic_route_config_providers_.erase(manager_identifier_);
-}
-
-void RdsRouteConfigSubscription::onConfigUpdate(
-    const std::vector<Envoy::Config::DecodedResourceRef>& resources,
-    const std::string& version_info) {
-  if (!validateUpdateSize(resources.size())) {
-    return;
+    const std::string& stat_prefix, Rds::RouteConfigProviderManager& route_config_provider_manager)
+    : Rds::RdsRouteConfigSubscription(
+          RouteConfigUpdatePtr(config_update),
+          std::unique_ptr<Envoy::Config::OpaqueResourceDecoder>(resource_decoder),
+          rds.config_source(), rds.route_config_name(), manager_identifier, factory_context,
+          stat_prefix + "rds.", "RDS", route_config_provider_manager),
+      config_update_info_(config_update) {}
+
+RdsRouteConfigSubscription::~RdsRouteConfigSubscription() { config_update_info_.release(); }
+
+void RdsRouteConfigSubscription::beforeProviderUpdate() {
+  if (config_update_info_->protobufConfigurationCast().has_vhds() &&
+      config_update_info_->vhdsConfigurationChanged()) {
+    std::unique_ptr<Init::ManagerImpl> noop_init_manager;
+    std::unique_ptr<Cleanup> resume_rds;
```

Comment:
Previously the cleanup was called after it went out of scope (after the `onConfigUpdate()` has finished), but now it is called after this method is done.
I'm not an expert here, but I think it defeats the purpose of the original code, or otherwise, the original code could have been much simpler, and would not have used the Cleanup object, no?

Reply:
Ok you are right, I change it to equivalent working.

Reply:
Done